dali/pipeline/operator/false_gpu_operator.h

```diff
@@ -0,0 +1,129 @@
+ *        the CPU outputs to device memory after Run.
+ *
+ * @tparam CPUOperator CPU operator class
+ */
+template <typename CPUOperator>
+class FalseGPUOperator : public Operator<GPUBackend> {
+ public:
+  explicit FalseGPUOperator(const OpSpec &spec)
+      : Operator<GPUBackend>(spec),
+        cpu_impl_(spec),
+        thread_pool_(num_threads_, spec.GetArgument<int>("device_id"), true /** set_affine */ ) {
+    cpu_ws_.SetThreadPool(&thread_pool_);
+  }
+  ~FalseGPUOperator() override = default;
+
+ protected:
+  bool CanInferOutputs() const override {
+    // To run Setup we need to first copy from device to host.
+    // To avoid delaying the Setup stage, we will do Setup and Run in one go (during Run)
+    return false;
+  }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const DeviceWorkspace &ws) {
+    return false;
+  }
+
+  void RunImpl(workspace_t<GPUBackend> &ws) override {
+    if (cpu_ws_.NumInput() == 0 && cpu_ws_.NumOutput() == 0) {
+      cpu_inputs_.resize(ws.NumInput());
+      for (int input_idx = 0; input_idx < ws.NumInput(); input_idx++) {
+        cpu_inputs_[input_idx] = std::make_shared<TensorVector<CPUBackend>>();
+        cpu_inputs_[input_idx]->set_pinned(true);
+        cpu_inputs_[input_idx]->set_order(AccessOrder::host());
+        cpu_ws_.AddInput(cpu_inputs_[input_idx]);
+      }
+      for (int output_idx = 0; output_idx < ws.NumOutput(); output_idx++) {
+        auto cpu_output = std::make_shared<TensorVector<CPUBackend>>();
+        cpu_output->set_pinned(true);
+        cpu_output->set_order(AccessOrder::host());
+        cpu_ws_.AddOutput(std::move(cpu_output));
+      }
+    } else {
+      // Number of inputs/outputs should not change after first iteration
+      assert(ws.NumInput() == cpu_ws_.NumInput());
+      assert(ws.NumOutput() == cpu_ws_.NumOutput());
+    }
+
+    for (int input_idx = 0; input_idx < ws.NumInput(); input_idx++) {
+      if (ws.InputIsType<GPUBackend>(0)) {
+        auto& gpu_input = ws.Input<GPUBackend>(input_idx);
+        cpu_inputs_[input_idx]->Copy(gpu_input);
+      } else {
+        // Some GPU operators might accept some CPU inputs (e.g. Slice)
+        auto& cpu_input = ws.Input<CPUBackend>(input_idx);
+        cpu_inputs_[input_idx]->ShareData(cpu_input);
+      }
+    }
+
+    output_desc_.clear();
+    if (cpu_impl_.Setup(output_desc_, cpu_ws_)) {
+      assert(static_cast<int>(output_desc_.size()) == cpu_ws_.NumOutput());
+      for (int output_idx = 0; output_idx < cpu_ws_.NumOutput(); output_idx++) {
+        auto &desc = output_desc_[output_idx];
+        cpu_ws_.template Output<CPUBackend>(output_idx).Resize(desc.shape, desc.type);
+      }
+    }
+
+    cpu_impl_.Run(cpu_ws_);
+
+    for (int output_idx = 0; output_idx < ws.NumOutput(); output_idx++) {
+      const auto& cpu_output = cpu_ws_.Output<CPUBackend>(output_idx);
+      ws.Output<GPUBackend>(output_idx).Copy(cpu_output, AccessOrder(ws.stream()));
```

Comment:
Better not to create AccessOrder in a loop - unless you know the device id.
AccessOrder(stream, device) is cheap
AccessOrder(stream) is much more expensive (needs to look up the device id by stream).